### ISSUE TITLE
Evenly distributed samples?

### ISSUE THREAD

**Issue by melodyjulia** [SciML/QuasiMonteCarlo.jl#69]
Please try the following case:

s1 = QuasiMonteCarlo.sample(25, [0.0], [1.0], LatinHypercubeSample())
sort(s1[1,:])’
s2 = QuasiMonteCarlo.sample(25, [0.0], [1.0], LatinHypercubeSample())
sort(s2[1,:])’
s3 = QuasiMonteCarlo.sample(25, [0.0], [1.0], LatinHypercubeSample())
sort(s3[1,:])’
s4 = QuasiMonteCarlo.sample(25, [0.0], [1.0], LatinHypercubeSample())
sort(s4[1,:])’

Output:
(sort(s1[1, :]))' = [0.02 0.06 0.1 0.13999999999999999 0.18 0.22 0.26 0.30000000000000004 0.34 0.38 0.42000000000000004 0.46 0.5 0.54 0.5800000000000001 0.62 0.66 0.7000000000000001 0.74 0.78 0.8200000000000001 0.86 0.9 0.9400000000000001 0.98]
(sort(s2[1, :]))' = [0.02 0.06 0.1 0.13999999999999999 0.18 0.22 0.26 0.30000000000000004 0.34 0.38 0.42000000000000004 0.46 0.5 0.54 0.5800000000000001 0.62 0.66 0.7000000000000001 0.74 0.78 0.8200000000000001 0.86 0.9 0.9400000000000001 0.98]
(sort(s3[1, :]))' = [0.02 0.06 0.1 0.13999999999999999 0.18 0.22 0.26 0.30000000000000004 0.34 0.38 0.42000000000000004 0.46 0.5 0.54 0.5800000000000001 0.62 0.66 0.7000000000000001 0.74 0.78 0.8200000000000001 0.86 0.9 0.9400000000000001 0.98]
(sort(s4[1, :]))' = [0.02 0.06 0.1 0.13999999999999999 0.18 0.22 0.26 0.30000000000000004 0.34 0.38 0.42000000000000004 0.46 0.5 0.54 0.5800000000000001 0.62 0.66 0.7000000000000001 0.74 0.78 0.8200000000000001 0.86 0.9 0.9400000000000001 0.98]

All samples are located at the centers of 25 evenly divided subregions: such as [0,0.04], [0.04, 0.08], ...., [0.96, 1.0].
Meanwhile, the results from four times sampling are same except their orders before sorting.

**Comment by ParadaCarleton** [comment]
Hi, that's correct. Latin Hypercube Samples are a random set of points, but stratified along the one-dimensional margins to make sure all points are evenly spaced out. Here's a picture:
![image](https://user-images.githubusercontent.com/71727937/214459584-a09e9a25-e606-4974-a0ee-dfdd9523ae11.png)


Could you explain what problem you're having?

**Comment by melodyjulia** [comment]
<img width="958" alt="Screen Shot 2023-01-26 at 10 05 13" src="https://user-images.githubusercontent.com/23096762/214742101-085e564a-c2fe-4796-8516-1cae1f09f39e.png">
<img width="554" alt="Screen Shot 2023-01-26 at 10 07 03" src="https://user-images.githubusercontent.com/23096762/214742119-1954f045-7bad-44bb-a2bc-d49ec340d9e7.png">

I compared the samples drawn by QuasiMonteCarlo.jl and R package lhs. The samples from QuasiMonteCarlo.jl are evenly distributed on the edges of evenly divided subregions. The samples from R packages "lhs" are randomly distributed in the evenly divided subregions. Please see the attached two figures.

**Comment by dmetivie** [comment]
Have a look at the definition in the package, [here](https://github.com/SciML/QuasiMonteCarlo.jl/blob/cc0c7f5f01eda7ba37739fb1673ae98bd482c871/src/LatinHypercube.jl#L10). Only the shuffling is random hence it explains the current behavior. 

LHS definition is 
`A Latin Hypercube is a point set with the property that every one-dimensional interval `(i/n, i+1/n)` contains exactly one point`
So the current behavior is ok with that (intervals are open on one side and close on the other). It is the bare minimum LHS in some sense.

What the R package is doing is a mystery (I tried to find the source code with no luck). 

Most probably it does something like 
```
function LHS(n, d)
    seq = ((1:n) - rand(n)) / n 
            # or ((1:n) .- rand()) / n 
            # or ((1:n) .- rand(1:n)/n) / n 
    return reduce(vcat, [shuffle(seq)' for _ in 1:d])
  end
```
The more randomness you add, the better your points will be. I think there is a whole literature (which I don't know) that does more elaborated LHS. 

Maybe more elaborate LHS methods could be a nice addition?

**Comment by ParadaCarleton** [comment]
> Maybe more elaborate LHS methods could be a nice addition?

They would be; some are already implemented in LHS.jl.

**Comment by ParadaCarleton** [comment]
> The more randomness you add, the better your points will be. I think there is a whole literature (which I don't know) that does more elaborated LHS.

Well, not necessarily, since more randomness means more variance. This implementation is supposed to be a centered LHS, i.e. the points are placed in the center of each interval. This reduces the variance slightly relative to placing points randomly within each stratum (by making sure the projections are perfectly uniform).

It does seem like there's a small bug though, in that points are apparently being placed at the ends of each interval, rather than in the center.

**Comment by melodyjulia** [comment]
It seems the current master version places the points in the center of each interval. The last released version v0.2.19 placed them at the ends of each interval.
